avalanche/training/plugins/generative_replay.py

```diff
@@ -0,0 +1,157 @@
+from avalanche.training.templates.supervised import SupervisedTemplate
+import torch
+
+
+class GenerativeReplayPlugin(SupervisedPlugin):
+    """
+    Experience generative replay plugin.
+
+    Updates the Dataloader of a strategy before training an experience
+    by sampling a generator model and weaving the replay data into
+    the original training data. 
+
+    The examples in the created mini-batch contain one part of the original data
+    and one part of generative data for each class 
+    that has been encountered so far.
+
+    In this version of the plugin the number of replay samples is 
+    increased with each new experience. Another way to implempent 
+    the algorithm is by weighting the loss function and give more 
+    importance to the replayed data as the number of experiences 
+    increases. This will be implemented as an option for the user soon.
+
+    :param batch_size: the size of the data batch. If set to `None`, it
+        will be set equal to the strategy's batch size.
+    :param batch_size_mem: the size of the memory batch. If
+        `task_balanced_dataloader` is set to True, it must be greater than or
+        equal to the number of tasks. If its value is set to `None`
+        (the default value), it will be automatically set equal to the
+        data batch size.
+    :param task_balanced_dataloader: if True, buffer data loaders will be
+            task-balanced, otherwise it will create a single dataloader for the
+            buffer samples.
+    :param untrained_solver: if True we assume this is the beginning of 
+        a continual learning task and add replay data only from the second 
+        experience onwards, otherwise we sample and add generative replay data
+        before training the first experience. Default to True.
+    """
+
+    def __init__(self, generator=None, mem_size: int = 200, 
+                 batch_size: int = None,
+                 batch_size_mem: int = None,
+                 task_balanced_dataloader: bool = False,
+                 untrained_solver: bool = True):
+        '''
+        Init.
+        '''
+        super().__init__()
+        self.mem_size = mem_size
+        self.batch_size = batch_size
+        self.batch_size_mem = batch_size_mem
+        self.task_balanced_dataloader = task_balanced_dataloader
+        self.generator_strategy = generator
+        if self.generator_strategy:
+            self.generator = generator.model
+        else: 
+            self.generator = None
+        self.untrained_solver = untrained_solver
+        self.model_is_generator = False
+
+    def before_training(self, strategy: "SupervisedTemplate", *args, **kwargs):
+        """Checks whether we are using a user defined external generator 
+        or we use the strategy's model as the generator. 
+        If the generator is None after initialization 
+        we assume that strategy.model is the generator."""
+        if not self.generator_strategy:
```

Comment:
Yes exactly: if no generative strategy is passed, we assume that the strategy our plugin was added to already has a model which is generative. E.g. this allows us to easily train a generator with generative replay by simply adding the plugin. In that case there would be no classifier. (Another scenario is that the classifier and generator are combined in a single model, as in this paper about [Generative replay with feedback connections](https://arxiv.org/abs/1809.10635)).

I agree that it can be confusing (a result of trying to combine all scenarios in a single plugin). I added another line in the doc string (4f5246f65d499df2e9015bbdd4d9fa2122c1f378) to refer to the example of training a generator, hoping this would make it clearer. Or do you have another suggestion?

Reply:
No it's fine, I read the code sequentially, so I didn't have a complete idea until the end. 
I still believe is a bit confusing, but I don't have a better idea at the moment. 
I really like the generality of the strategy, and overall the pros of this implementation greatly outshine a bit of confusion in this part :wink: